lib/bakeAmbientOcclusion.js

```diff
@@ -0,0 +1,586 @@
+    }
+
+    // Build a hash of materials, textures, and images we've seen so far to ensure uniqueness
+    var parameters = {
+        materialsSeen: {},
+        texturesSeen: {},
+        imagesSeen: {},
+        exampleMaterialID: exampleMaterialID,
+        exampleTextureID: exampleTextureID,
+        exampleImageID: exampleImageID,
+        resolution: resolution,
+        gltf: gltf,
+        raytracerScene: raytracerScene
+    };
+
+    // Bake AO for each primitive in the scene
+    NodeHelpers.forEachPrimitiveInScene(gltf, scene, addAoToImage, parameters);
+}
+
+function addAoToImage(primitive, meshPrimitiveID, parameters) {
+    // Enforce material/texture/image uniqueness
+    var gltf = parameters.gltf;
+    var diffuseImage = ensureImageUniqueness(gltf, primitive, meshPrimitiveID, parameters);
+    diffuseImage.extras._pipeline.imageChanged = true;
+    var diffuseImageJimp = diffuseImage.extras._pipeline.jimpImage;
+    var goalResolution = diffuseImageJimp.bitmap.width;
+
+    // Post process the AO
+    var jimpAO = gltf.extras._pipeline.jimpScratch;
+    var aoBuffer = parameters.raytracerScene.aoBufferByPrimitive[meshPrimitiveID];
+    postProcessAO(aoBuffer, parameters.resolution, goalResolution, jimpAO);
+
+    // Modify the diffuse image with AO
+    for (var x = 0; x < goalResolution; x++) {
+        for (var y = 0; y < goalResolution; y++) {
+            var idx = (goalResolution * y + x) * 4;
+            var aoValue = 1.0 - (jimpAO.bitmap.data[idx + 3] / 255.0);
+
+            // darken each channel by the ao value
+            diffuseImageJimp.bitmap.data[idx] *= aoValue;
+            diffuseImageJimp.bitmap.data[idx + 1] *= aoValue;
+            diffuseImageJimp.bitmap.data[idx + 2] *= aoValue;
+        }
+    }
+}
+
+function postProcessAO(aoBuffer, dataResolution, goalResolution, jimpAO) {
+    // Copy the data over to the jimp
+    var value = 0.0;
+    jimpAO.resize(dataResolution, dataResolution);
+    for (var x = 0; x < dataResolution; x++) {
+        for (var y = 0; y < dataResolution; y++) {
+            var dataIdx = dataResolution * y + x;
+            value = 255.0 * (aoBuffer.samples[dataIdx] / aoBuffer.count[dataIdx]);
+            jimpAO.bitmap.data[dataIdx * 4 + 3] = value;
+        }
+    }
+    // Resize the data to match the goal resolution
+    jimpAO.resize(goalResolution, goalResolution, Jimp.RESIZE_BEZIER);
+}
+
+function cloneAndSetup(newMateralID, newTextureID, newImageID,
+                      oldMateralID, oldTextureID, oldImageID,
+                      materials, textures, images) {
+    var newImage;
+    var newTexture;
+    var newMaterial;
+
+    if (defined(newImageID)) {
+        var oldImage = images[oldImageID];
+        newImage = clone(oldImage);
+        newImage.extras._pipeline.jimpImage = oldImage.extras._pipeline.jimpImage.clone();
+        images[newImageID] = newImage;
+    }
+    if (defined(newTextureID)) {
```

Comment:
It is better style to move `if (defined(newMateralID)) {` into this block and drop the `else` block with the `return`.